### ISSUE TITLE
Bundler 2.2.1 - bundle package fails

### ISSUE THREAD

**Issue by sterankin** [rubygems/rubygems#4153]
Our CI build broke today with the below error.  Its seems the Docker image is install bundler 2.2.1 and when we run docker package, it generates the following error report:

I was able to fix by installing bundler 2.1.4 and using that in my CI build, but I am hoping someone can help with the below.

Ruby 2.5.8

**What did you do?**

I ran the command `/usr/local/bundle/bin/bundle package`

**What did you expect to happen?**

I expected Bundler to bundle package as normal

 **What happened instead?**

Instead, I got the error below.

**Have you tried any solutions posted on similar issues in our issue tracker, stack overflow, or google?**
No, I downgraded bundler to 2.1.4 to get it to work

## Backtrace

```
RuntimeError: Specs already loaded
/usr/local/bundle/gems/bundler-2.2.1/lib/bundler/definition.rb:160:in `resolve_remotely!'
usr/local/bundle/gems/bundler-2.2.1/lib/bundler/installer.rb:306:in `resolve_if_needed'
 /usr/local/bundle/gems/bundler-2.2.1/lib/bundler/installer.rb:83:in `block in run'
/usr/local/bundle/gems/bundler-2.2.1/lib/bundler/process_lock.rb:12:in `block in lock'
/usr/local/bundle/gems/bundler-2.2.1/lib/bundler/process_lock.rb:9:in `open'
/usr/local/bundle/gems/bundler-2.2.1/lib/bundler/process_lock.rb:9:in `lock'
 /usr/local/bundle/gems/bundler-2.2.1/lib/bundler/installer.rb:72:in `run'
/usr/local/bundle/gems/bundler-2.2.1/lib/bundler/installer.rb:24:in `install'
 /usr/local/bundle/gems/bundler-2.2.1/lib/bundler/cli/install.rb:64:in `run'
/usr/local/bundle/gems/bundler-2.2.1/lib/bundler/cli/cache.rb:33:in `install'
/usr/local/bundle/gems/bundler-2.2.1/lib/bundler/cli/cache.rb:17:in `run'
/usr/local/bundle/gems/bundler-2.2.1/lib/bundler/cli.rb:482:in `cache'
/usr/local/bundle/gems/bundler-2.2.1/lib/bundler/vendor/thor/lib/thor/command.rb:27:in `run'
/usr/local/bundle/gems/bundler-2.2.1/lib/bundler/vendor/thor/lib/thor/invocation.rb:127:in `invoke_command'
/usr/local/bundle/gems/bundler-2.2.1/lib/bundler/vendor/thor/lib/thor.rb:392:in `dispatch'
/usr/local/bundle/gems/bundler-2.2.1/lib/bundler/cli.rb:30:in `dispatch'
/usr/local/bundle/gems/bundler-2.2.1/lib/bundler/vendor/thor/lib/thor/base.rb:485:in `start'
/usr/local/bundle/gems/bundler-2.2.1/lib/bundler/cli.rb:24:in `start'
/usr/local/bundle/gems/bundler-2.2.1/exe/bundle:49:in `block in <top (required)>'
/usr/local/bundle/gems/bundler-2.2.1/lib/bundler/friendly_errors.rb:130:in `with_friendly_errors'
/usr/local/bundle/gems/bundler-2.2.1/exe/bundle:37:in `<top (required)>'
 /usr/local/bundle/bin/bundle:23:in `load'
 /usr/local/bundle/bin/bundle:23:in `<top (required)>'
/usr/local/bundle/gems/bundler-2.2.1/lib/bundler/cli/exec.rb:63:in `load'
 /usr/local/bundle/gems/bundler-2.2.1/lib/bundler/cli/exec.rb:63:in `kernel_load'
/usr/local/bundle/gems/bundler-2.2.1/lib/bundler/cli/exec.rb:28:in `run'
/usr/local/bundle/gems/bundler-2.2.1/lib/bundler/cli.rb:497:in `exec'
/usr/local/bundle/gems/bundler-2.2.1/lib/bundler/vendor/thor/lib/thor/command.rb:27:in `run'
/usr/local/bundle/gems/bundler-2.2.1/lib/bundler/vendor/thor/lib/thor/invocation.rb:127:in `invoke_command'
/usr/local/bundle/gems/bundler-2.2.1/lib/bundler/vendor/thor/lib/thor.rb:392:in `dispatch'
/usr/local/bundle/gems/bundler-2.2.1/lib/bundler/cli.rb:30:in `dispatch'
/usr/local/bundle/gems/bundler-2.2.1/lib/bundler/vendor/thor/lib/thor/base.rb:485:in `start'
/usr/local/bundle/gems/bundler-2.2.1/lib/bundler/cli.rb:24:in `start'
 /usr/local/bundle/gems/bundler-2.2.1/exe/bundle:49:in `block in <top (required)>'
/usr/local/bundle/gems/bundler-2.2.1/lib/bundler/friendly_errors.rb:130:in `with_friendly_errors'
/usr/local/bundle/gems/bundler-2.2.1/exe/bundle:37:in `<top (required)>'
/usr/local/bundle/bin/bundle:23:in `load'
 /usr/local/bundle/bin/bundle:23:in `<main>'
```

 ## Environment

```
Bundler       2.2.1
Platforms   ruby, x86_64-linux
Ruby          2.5.8p224 (2020-03-31 revision 67882) [x86_64-linux]
Full Path   /usr/local/bin/ruby
Config Dir  /usr/local/etc
RubyGems      3.0.3
Gem Home    /usr/local/bundle
Gem Path    /root/.gem/ruby/2.5.0:/usr/local/lib/ruby/gems/2.5.0:/usr/local/bundle
User Home   /root
User Path   /root/.gem/ruby/2.5.0
Bin Dir     /usr/local/bundle/bin
OpenSSL
Compiled    OpenSSL 1.1.1d  10 Sep 2019
Loaded      OpenSSL 1.1.1d  10 Sep 2019
Cert File   /usr/lib/ssl/cert.pem
Cert Dir    /usr/lib/ssl/certs
Tools
Git         2.20.1
RVM         not installed
rbenv       not installed
chruby      not installed
```

## Bundler Build Metadata

```
Built At          2020-12-14
Git SHA           b98d6b2035
Released Version  true
```

## Bundler settings
  
```
jobs
Set for your local app (/usr/local/bundle/config): 4
cache_all
You have not configured a value for `cache_all`
app_config
Set via BUNDLE_APP_CONFIG: "/usr/local/bundle"
silence_root_warning
Set via BUNDLE_SILENCE_ROOT_WARNING: true
````

**Comment by deivid-rodriguez** [comment]
Hi @sterankin! Could you trim all the timestamps and add some formatting to the error, please? It's very hard to read. Here's a markdown guide: https://guides.github.com/features/mastering-markdown/.

Also, could you provide me with a set of steps I can run on my computer in order to reproduce the error?

Thanks.

**Comment by sterankin** [comment]
I have been trying to replicate this locally using Docker in the same manner as our CI build and I am not having any luck replicating the issue.

Put this on hold until I can recreate.

**Comment by deivid-rodriguez** [comment]
Can you also share the `Gemfile` and `Gemfile.lock`?

**Comment by sterankin** [comment]
Gemfile:

```
source 'https://rubygems.org'

ruby '2.5.8'

gem 'rails', '~> 5.0.7.1'

# Use jquery as the JavaScript library
gem 'jquery-rails', '4.3.5'
gem 'jquery-turbolinks'

gem 'haml'
gem 'bootstrap-sass', '3.3.5'
gem 'sass-rails', '5.0.5'
gem 'coffee-rails', ' 4.2.2'
gem 'uglifier', '>= 1.0.3'
gem "render_async"

#For REST API integration/consumption
gem 'httparty', '>= 0.12.0'

# Build JSON APIs with ease. Read more: https://github.com/rails/jbuilder
gem 'jbuilder', '~> 2.0'

# Use swagger for easy and interactive API documentation
gem 'swagger-ui_rails'
gem 'swagger-docs'

gem 'tzinfo', '1.2.2'

group :doc do
  # bundle exec rake doc:rails generates the API under doc/api.
  gem 'sdoc', '0.4.0', require: false
end


platforms :ruby, :mswin, :mingw do
  gem 'sqlite3', '1.3.10'
  gem 'tzinfo-data', '1.2014.6'

  group :development, :test do
    gem 'better_errors'
    #gem 'bullet', '4.8.0'
    gem 'brakeman', '2.3.1'
    gem 'byebug'
    gem 'binding_of_caller', '0.7.2'
    gem 'jasmine', '~> 2.0.0'
    gem 'wdm', '>= 0.1.0' if RbConfig::CONFIG['target_os'] =~ /mswin|mingw|cygwin/i
  end
end

group :development, :test do
  gem 'minitest', '5.4.0'
  gem 'rspec-rails', '3.6'
  gem 'shoulda-matchers', '3.1.2'
  gem "factory_bot_rails", '4.10.0'
  gem 'simplecov', '0.8.2'
  gem 'simplecov-bamboo', '0.1.0'
  gem 'dotenv-rails'
  gem 'pry-rescue'
  gem 'pry-stack_explorer'
end

group :development do
  gem 'guard-rspec', '4.7.3', require: false

  platforms :mingw, :x64_mingw, :mswin do
    gem 'ffi' # Required by overcommit when using windows
  end
end

group :test do
  gem 'faker', '1.2.0'
  gem 'launchy', '2.4.2'
  gem 'database_cleaner', '1.2.0'
  gem 'capybara', '3.15.1'
  gem 'rspec_junit_formatter'
  gem 'rack_session_access', '0.1.1'
  gem 'rails-controller-testing'
end

gem 'kaminari'
gem 'ruby-ntlm', '0.0.1'
gem 'nokogiri', '1.10.9'
gem 'rack-mini-profiler', '0.9.0', require: false
gem 'annotate', '~> 2.6.5'

#Gem for state machine
gem 'statesman', '2.0.1'

#Gems added for ADFS/SAML integration
gem 'omniauth-ldap', '~> 1.0'
gem 'omniauth-saml-rstr', '0.1.0'
gem 'omniauth-oauth2', '~> 1.0'

gem 'certified'

# Gems for flipper ActiveRecord
gem 'flipper'
gem 'flipper-active_record'


gem 'sidekiq', '5.2.8'
gem 'sidekiq-scheduler', '3.0.0'
gem 'puma'

gem 'pry'

gem 'cityhash'
gem 'sprockets', '3.7.2'
```


Gemfile lock

```
GEM
  remote: https://rubygems.org/
  specs:
    actioncable (5.0.7.2)
      actionpack (= 5.0.7.2)
      nio4r (>= 1.2, < 3.0)
      websocket-driver (~> 0.6.1)
    actionmailer (5.0.7.2)
      actionpack (= 5.0.7.2)
      actionview (= 5.0.7.2)
      activejob (= 5.0.7.2)
      mail (~> 2.5, >= 2.5.4)
      rails-dom-testing (~> 2.0)
    actionpack (5.0.7.2)
      actionview (= 5.0.7.2)
      activesupport (= 5.0.7.2)
      rack (~> 2.0)
      rack-test (~> 0.6.3)
      rails-dom-testing (~> 2.0)
      rails-html-sanitizer (~> 1.0, >= 1.0.2)
    actionview (5.0.7.2)
      activesupport (= 5.0.7.2)
      builder (~> 3.1)
      erubis (~> 2.7.0)
      rails-dom-testing (~> 2.0)
      rails-html-sanitizer (~> 1.0, >= 1.0.3)
    activejob (5.0.7.2)
      activesupport (= 5.0.7.2)
      globalid (>= 0.3.6)
    activemodel (5.0.7.2)
      activesupport (= 5.0.7.2)
    activerecord (5.0.7.2)
      activemodel (= 5.0.7.2)
      activesupport (= 5.0.7.2)
      arel (~> 7.0)
    activesupport (5.0.7.2)
      concurrent-ruby (~> 1.0, >= 1.0.2)
      i18n (>= 0.7, < 2)
      minitest (~> 5.1)
      tzinfo (~> 1.1)
    addressable (2.7.0)
      public_suffix (>= 2.0.2, < 5.0)
    annotate (2.6.7)
      activerecord (>= 2.3.0)
      rake (~> 10.4.2, >= 10.4.2)
    arel (7.1.4)
    autoprefixer-rails (10.1.0.0)
      execjs
    better_errors (2.9.1)
      coderay (>= 1.0.0)
      erubi (>= 1.0.0)
      rack (>= 0.9.0)
    binding_of_caller (0.7.2)
      debug_inspector (>= 0.0.1)
    bootstrap-sass (3.3.5)
      autoprefixer-rails (>= 5.0.0.1)
      sass (>= 3.2.19)
    brakeman (2.3.1)
      erubis (~> 2.6)
      fastercsv (~> 1.5)
      haml (>= 3.0, < 5.0)
      highline (~> 1.6.20)
      multi_json (~> 1.2)
      ruby2ruby (~> 2.0.5)
      ruby_parser (~> 3.2.2)
      sass (~> 3.0)
      slim (>= 1.3.6, < 3.0)
      terminal-table (~> 1.4)
    builder (3.2.4)
    byebug (11.1.3)
    capybara (3.15.1)
      addressable
      mini_mime (>= 0.1.3)
      nokogiri (~> 1.8)
      rack (>= 1.6.0)
      rack-test (>= 0.6.3)
      regexp_parser (~> 1.2)
      xpath (~> 3.2)
    certified (1.0.0)
    cityhash (0.9.0)
    coderay (1.1.3)
    coffee-rails (4.2.2)
      coffee-script (>= 2.2.0)
      railties (>= 4.0.0)
    coffee-script (2.4.1)
      coffee-script-source
      execjs
    coffee-script-source (1.12.2)
    concurrent-ruby (1.1.7)
    connection_pool (2.2.3)
    crass (1.0.6)
    database_cleaner (1.2.0)
    debug_inspector (0.0.3)
    diff-lcs (1.4.4)
    docile (1.1.5)
    dotenv (2.7.6)
    dotenv-rails (2.7.6)
      dotenv (= 2.7.6)
      railties (>= 3.2)
    erubi (1.10.0)
    erubis (2.7.0)
    et-orbi (1.2.4)
      tzinfo
    execjs (2.7.0)
    factory_bot (4.10.0)
      activesupport (>= 3.0.0)
    factory_bot_rails (4.10.0)
      factory_bot (~> 4.10.0)
      railties (>= 3.0.0)
    faker (1.2.0)
      i18n (~> 0.5)
    faraday (1.1.0)
      multipart-post (>= 1.2, < 3)
      ruby2_keywords
    fastercsv (1.5.5)
    ffi (1.13.1)
    flipper (0.19.1)
    flipper-active_record (0.19.1)
      activerecord (>= 5.0, < 7)
      flipper (~> 0.19.1)
    formatador (0.2.5)
    fugit (1.4.1)
      et-orbi (~> 1.1, >= 1.1.8)
      raabro (~> 1.4)
    globalid (0.4.2)
      activesupport (>= 4.2.0)
    guard (2.16.2)
      formatador (>= 0.2.4)
      listen (>= 2.7, < 4.0)
      lumberjack (>= 1.0.12, < 2.0)
      nenv (~> 0.1)
      notiffany (~> 0.0)
      pry (>= 0.9.12)
      shellany (~> 0.0)
      thor (>= 0.18.1)
    guard-compat (1.2.1)
    guard-rspec (4.7.3)
      guard (~> 2.1)
      guard-compat (~> 1.1)
      rspec (>= 2.99.0, < 4.0)
    haml (4.0.7)
      tilt
    hashie (4.1.0)
    highline (1.6.21)
    httparty (0.18.1)
      mime-types (~> 3.0)
      multi_xml (>= 0.5.2)
    i18n (0.9.5)
      concurrent-ruby (~> 1.0)
    interception (0.5)
    jasmine (2.0.3)
      jasmine-core (~> 2.0.0)
      phantomjs
      rack (>= 1.2.1)
      rake
    jasmine-core (2.0.2)
    jbuilder (2.10.1)
      activesupport (>= 5.0.0)
    jquery-rails (4.3.5)
      rails-dom-testing (>= 1, < 3)
      railties (>= 4.2.0)
      thor (>= 0.14, < 2.0)
    jquery-turbolinks (2.1.0)
      railties (>= 3.1.0)
      turbolinks
    json (1.8.6)
    jwt (2.2.2)
    kaminari (1.2.1)
      activesupport (>= 4.1.0)
      kaminari-actionview (= 1.2.1)
      kaminari-activerecord (= 1.2.1)
      kaminari-core (= 1.2.1)
    kaminari-actionview (1.2.1)
      actionview
      kaminari-core (= 1.2.1)
    kaminari-activerecord (1.2.1)
      activerecord
      kaminari-core (= 1.2.1)
    kaminari-core (1.2.1)
    launchy (2.4.2)
      addressable (~> 2.3)
    listen (3.3.3)
      rb-fsevent (~> 0.10, >= 0.10.3)
      rb-inotify (~> 0.9, >= 0.9.10)
    loofah (2.8.0)
      crass (~> 1.0.2)
      nokogiri (>= 1.5.9)
    lumberjack (1.2.8)
    macaddr (1.7.2)
      systemu (~> 2.6.5)
    mail (2.7.1)
      mini_mime (>= 0.1.1)
    method_source (1.0.0)
    mime-types (3.3.1)
      mime-types-data (~> 3.2015)
    mime-types-data (3.2020.1104)
    mini_mime (1.0.2)
    mini_portile2 (2.4.0)
    minitest (5.4.0)
    multi_json (1.15.0)
    multi_xml (0.6.0)
    multipart-post (2.1.1)
    nenv (0.3.0)
    net-ldap (0.17.0)
    nio4r (2.5.4)
    nokogiri (1.10.9)
      mini_portile2 (~> 2.4.0)
    notiffany (0.1.3)
      nenv (~> 0.1)
      shellany (~> 0.0)
    oauth2 (1.4.4)
      faraday (>= 0.8, < 2.0)
      jwt (>= 1.0, < 3.0)
      multi_json (~> 1.3)
      multi_xml (~> 0.5)
      rack (>= 1.2, < 3)
    omniauth (1.9.1)
      hashie (>= 3.4.6)
      rack (>= 1.6.2, < 3)
    omniauth-ldap (1.0.5)
      net-ldap (~> 0.12)
      omniauth (~> 1.0)
      pyu-ruby-sasl (~> 0.0.3.2)
      rubyntlm (~> 0.3.4)
    omniauth-oauth2 (1.7.0)
      oauth2 (~> 1.4)
      omniauth (~> 1.9)
    omniauth-saml-rstr (0.1.0)
      omniauth (~> 1.0)
      uuid (~> 2.3)
      xmlcanonicalizer (= 0.1.1)
    phantomjs (2.1.1.0)
    pry (0.13.1)
      coderay (~> 1.1)
      method_source (~> 1.0)
    pry-rescue (1.5.2)
      interception (>= 0.5)
      pry (>= 0.12.0)
    pry-stack_explorer (0.4.12)
      binding_of_caller (~> 0.7)
      pry (~> 0.13)
    public_suffix (4.0.6)
    puma (5.1.1)
      nio4r (~> 2.0)
    pyu-ruby-sasl (0.0.3.3)
    raabro (1.4.0)
    rack (2.0.9)
    rack-mini-profiler (0.9.0)
      rack (>= 1.1.3)
    rack-protection (2.1.0)
      rack
    rack-test (0.6.3)
      rack (>= 1.0)
    rack_session_access (0.1.1)
      builder (>= 2.0.0)
      rack (>= 1.0.0)
    rails (5.0.7.2)
      actioncable (= 5.0.7.2)
      actionmailer (= 5.0.7.2)
      actionpack (= 5.0.7.2)
      actionview (= 5.0.7.2)
      activejob (= 5.0.7.2)
      activemodel (= 5.0.7.2)
      activerecord (= 5.0.7.2)
      activesupport (= 5.0.7.2)
      bundler (>= 1.3.0)
      railties (= 5.0.7.2)
      sprockets-rails (>= 2.0.0)
    rails-controller-testing (1.0.5)
      actionpack (>= 5.0.1.rc1)
      actionview (>= 5.0.1.rc1)
      activesupport (>= 5.0.1.rc1)
    rails-dom-testing (2.0.3)
      activesupport (>= 4.2.0)
      nokogiri (>= 1.6)
    rails-html-sanitizer (1.3.0)
      loofah (~> 2.3)
    railties (5.0.7.2)
      actionpack (= 5.0.7.2)
      activesupport (= 5.0.7.2)
      method_source
      rake (>= 0.8.7)
      thor (>= 0.18.1, < 2.0)
    rake (10.4.2)
    rb-fsevent (0.10.4)
    rb-inotify (0.10.1)
      ffi (~> 1.0)
    rdoc (4.3.0)
    redis (4.2.5)
    regexp_parser (1.8.2)
    render_async (2.1.8)
    rspec (3.6.0)
      rspec-core (~> 3.6.0)
      rspec-expectations (~> 3.6.0)
      rspec-mocks (~> 3.6.0)
    rspec-core (3.6.0)
      rspec-support (~> 3.6.0)
    rspec-expectations (3.6.0)
      diff-lcs (>= 1.2.0, < 2.0)
      rspec-support (~> 3.6.0)
    rspec-mocks (3.6.0)
      diff-lcs (>= 1.2.0, < 2.0)
      rspec-support (~> 3.6.0)
    rspec-rails (3.6.0)
      actionpack (>= 3.0)
      activesupport (>= 3.0)
      railties (>= 3.0)
      rspec-core (~> 3.6.0)
      rspec-expectations (~> 3.6.0)
      rspec-mocks (~> 3.6.0)
      rspec-support (~> 3.6.0)
    rspec-support (3.6.0)
    rspec_junit_formatter (0.4.1)
      rspec-core (>= 2, < 4, != 2.12.0)
    ruby-ntlm (0.0.1)
    ruby2_keywords (0.0.2)
    ruby2ruby (2.0.8)
      ruby_parser (~> 3.1)
      sexp_processor (~> 4.0)
    ruby_parser (3.2.2)
      sexp_processor (~> 4.1)
    rubyntlm (0.3.4)
    rufus-scheduler (3.6.0)
      fugit (~> 1.1, >= 1.1.6)
    sass (3.7.4)
      sass-listen (~> 4.0.0)
    sass-listen (4.0.0)
      rb-fsevent (~> 0.9, >= 0.9.4)
      rb-inotify (~> 0.9, >= 0.9.7)
    sass-rails (5.0.5)
      railties (>= 4.0.0, < 6)
      sass (~> 3.1)
      sprockets (>= 2.8, < 4.0)
      sprockets-rails (>= 2.0, < 4.0)
      tilt (>= 1.1, < 3)
    sdoc (0.4.0)
      json (~> 1.8)
      rdoc (~> 4.0, < 5.0)
    sexp_processor (4.15.1)
    shellany (0.0.1)
    shoulda-matchers (3.1.2)
      activesupport (>= 4.0.0)
    sidekiq (5.2.8)
      connection_pool (~> 2.2, >= 2.2.2)
      rack (< 2.1.0)
      rack-protection (>= 1.5.0)
      redis (>= 3.3.5, < 5)
    sidekiq-scheduler (3.0.0)
      redis (>= 3, < 5)
      rufus-scheduler (~> 3.2)
      sidekiq (>= 3)
      tilt (>= 1.4.0)
    simplecov (0.8.2)
      docile (~> 1.1.0)
      multi_json
      simplecov-html (~> 0.8.0)
    simplecov-bamboo (0.1.0)
      nokogiri
      simplecov
      simplecov-html
    simplecov-html (0.8.0)
    slim (2.1.0)
      temple (~> 0.6.9)
      tilt (>= 1.3.3, < 2.1)
    sprockets (3.7.2)
      concurrent-ruby (~> 1.0)
      rack (> 1, < 3)
    sprockets-rails (3.2.2)
      actionpack (>= 4.0)
      activesupport (>= 4.0)
      sprockets (>= 3.0.0)
    sqlite3 (1.3.10)
    statesman (2.0.1)
    swagger-docs (0.2.9)
      activesupport (>= 3)
      rails (>= 3)
    swagger-ui_rails (0.1.7)
    systemu (2.6.5)
    temple (0.6.10)
    terminal-table (1.8.0)
      unicode-display_width (~> 1.1, >= 1.1.1)
    thor (1.0.1)
    thread_safe (0.3.6)
    tilt (2.0.10)
    turbolinks (5.2.1)
      turbolinks-source (~> 5.2)
    turbolinks-source (5.2.0)
    tzinfo (1.2.2)
      thread_safe (~> 0.1)
    tzinfo-data (1.2014.6)
      tzinfo (>= 1.0.0)
    uglifier (4.2.0)
      execjs (>= 0.3.0, < 3)
    unicode-display_width (1.7.0)
    uuid (2.3.9)
      macaddr (~> 1.0)
    websocket-driver (0.6.5)
      websocket-extensions (>= 0.1.0)
    websocket-extensions (0.1.5)
    xmlcanonicalizer (0.1.1)
    xpath (3.2.0)
      nokogiri (~> 1.8)

PLATFORMS
  ruby

DEPENDENCIES
  annotate (~> 2.6.5)
  better_errors
  binding_of_caller (= 0.7.2)
  bootstrap-sass (= 3.3.5)
  brakeman (= 2.3.1)
  byebug
  capybara (= 3.15.1)
  certified
  cityhash
  coffee-rails (= 4.2.2)
  database_cleaner (= 1.2.0)
  dotenv-rails
  factory_bot_rails (= 4.10.0)
  faker (= 1.2.0)
  ffi
  flipper
  flipper-active_record
  guard-rspec (= 4.7.3)
  haml
  httparty (>= 0.12.0)
  jasmine (~> 2.0.0)
  jbuilder (~> 2.0)
  jquery-rails (= 4.3.5)
  jquery-turbolinks
  kaminari
  launchy (= 2.4.2)
  minitest (= 5.4.0)
  nokogiri (= 1.10.9)
  omniauth-ldap (~> 1.0)
  omniauth-oauth2 (~> 1.0)
  omniauth-saml-rstr (= 0.1.0)
  pry
  pry-rescue
  pry-stack_explorer
  puma
  rack-mini-profiler (= 0.9.0)
  rack_session_access (= 0.1.1)
  rails (~> 5.0.7.1)
  rails-controller-testing
  render_async
  rspec-rails (= 3.6)
  rspec_junit_formatter
  ruby-ntlm (= 0.0.1)
  sass-rails (= 5.0.5)
  sdoc (= 0.4.0)
  shoulda-matchers (= 3.1.2)
  sidekiq (= 5.2.8)
  sidekiq-scheduler (= 3.0.0)
  simplecov (= 0.8.2)
  simplecov-bamboo (= 0.1.0)
  sprockets (= 3.7.2)
  sqlite3 (= 1.3.10)
  statesman (= 2.0.1)
  swagger-docs
  swagger-ui_rails
  tzinfo (= 1.2.2)
  tzinfo-data (= 1.2014.6)
  uglifier (>= 1.0.3)

RUBY VERSION
   ruby 2.5.8p224

BUNDLED WITH
   2.1.4
```

**Comment by deivid-rodriguez** [comment]
As expected, I can't repro this. I'll put this on hold until you can recreate it!

**Comment by jeremy-ebler-vineti** [comment]
Sorry I don't have bandwidth to help debug, but I hit the same issue with `2.2.1`. Rolling back to `2.1.4` worked for me. I didn't test `2.2.0`

**Comment by gradybarrett** [comment]
@deivid-rodriguez @sterankin I think we can chalk this one up to the Dockerfile implementation.

tl/dr: Keep your versions of bundler in sync in your lock file and image.

It's probably not a good idea to rely on the image build to upgrade the lock file. If you wanted to get fancy you could take extra steps to preserve the updated lock file when copying in app code, but I don't think that's desirable.

Here's a simplified version of what we were doing in our Dockerfile that was causing problems:

```dockerfile
FROM ruby:2.5.8

WORKDIR /tmp

RUN gem install bundler # installs bundler 2.2.2

ADD Gemfile .
ADD Gemfile.lock . # Old lock file bundled using 2.1.4

RUN bundle install --jobs 4 # Upgrades the copied in lock file to 2.2.2 and installs gems, great! Or maybe not...

WORKDIR /usr/src/app
# Copy in the old lock file (using 2.1.4 w/o platforms declared)...whoops...
# The image was built using 2.2.2 but we reintroduced the old lock file, which is then used to run
# all subsequent commands in spun up containers. In our case, that was a `bundle exec bundle package`,
# which exploded because of the environment mismatch.
ADD . /usr/src/app
```

**Comment by deivid-rodriguez** [comment]
@gradybarrett Thanks!

That's very useful information, I think I might find a way to repro with that information. Maybe things should be done differently in the `Dockerfile`, what you explain sounds most likely like not a good idea, indeed.

However, even if we can't autoremediate things on the fly, and consider an error fine for this situation, in my opinion the current error being thrown is unacceptable because it doesn't give any useful information about the underlying problem.

I'll try to repro and improve this.

**Comment by deivid-rodriguez** [comment]
Unfortunately I still can't repro... :(

I tried building a container using 2.2.2, but keeping the old lockfile so that when it runs it uses that, but bundler will automatically update it on `bundle exec` when running the container and run the `bundle package` command properly. There's something more to this that we're missing.

I'm totally up for getting to the bottom of this, but I need some help to reproduce the issue and be able to understand what's going on.

**Comment by sterankin** [comment]
I was able to reproduce just now, its a bit convoluted - so bear with me.

1. Have a rails app and ensure the gems are bundled with bundler 2.1.4.  Ensure the rails app has a Docker file in the root dir from the comments above.
2. In the root directory of your rails app, create a subfolder, e.g. call it _myapp_temp_dir_ This is to mimic what is actually happening in our CI build.  Copy the **entire** Rails project into this subdirectory.  So now you have 2 copies of the rails app.
3. Run the following Docker build command,  run the command in the root directory of the rails app (not the sub directory you created)
`docker build --force-rm=true -t my_app ${PWD}/myapp_temp_dir`

4. This should build and do a bundle install using bundler 2.2.1
5. Now run the following docker run command:
`docker run --rm -v ${PWD}/myapp_temp_dir:/usr/src/app --rm my_app:latest  bash -c "bundle exec bundle package"`

This should generate the error.

It seems that the lockfile in the docker working subdirectory (_myapp_temp_dir_ above) is correctly a 2.2.1 version after the _docker build/bundle install.  But in our CI build, we are not copying across the working subdirectory, we are copying across the top level folder.  So when the bundle package runs it is using bundler version 2.2.1 on an older 2.1.4 lockfile from the top level dir.

**Comment by deivid-rodriguez** [comment]
Nice, are you able to publish a sample repository with these setup? Just to save me recreating everything? :)

**Comment by sterankin** [comment]
Here you go, clone this project.

https://github.com/sterankin/test-rails-docker-bundler

Notice that is has a subfolder called test_app which is a copy of the top level rails app, this mimics what is happening on our CI build.

To reproduce the error:

1. Ensure docker is running
2. Open a terminal and `cd` into the root folder
3. Run `docker build --force-rm=true --no-cache -t test_app  ${PWD}/test_app `
4. Run `docker run --rm -v ${PWD}/test_app:/usr/src/app --rm test_app:latest  bash -c "bundle exec bundle package"`

You should get the error.

**Comment by deivid-rodriguez** [comment]
Woooooow, nice, thanks so much! That's half of the way of fixing this!

I can't pull the base image though, `repo.forge.lmig.com` seems inaccessible for me.

**Comment by sterankin** [comment]
Oops, change the Dockerfile to:

`FROM ruby:2.5.8`

And you should be all set

**Comment by deivid-rodriguez** [comment]
And reproduced!!! Thanks! I'll work on this on Monday.

**Comment by gradybarrett** [comment]
For what it's worth, this doesn't appear to happen when using ruby 2.7.1.

Here's a simpler repro in case it's useful for comparison. It uses only a Gemfile and lock file with a single dependency.

https://github.com/gradybarrett/bundler-error-repro

**Comment by deivid-rodriguez** [comment]
Thank you everyone for the repros! I created #4175 that should fix this.

By the way, may I ask why you're running `bundle exec bundle`? It seems unnecessary, `bundle` should be enough. Is it to get the version of `bundler` locked in the lockfile? It's that's the reason, this should be the default if the version is installed, and we plan to make `bundle install` install that automatically.

**Comment by MehrshadMohammadi** [comment]
> Oops, change the Dockerfile to:
> 
> `FROM ruby:2.5.8`
> 
> And you should be all set

Hi ... 
this link -> https://github.com/sterankin/test-rails-docker-bundler
Only works on version 2.5.8 ?
I am using version ruby 3.0.0p0 but I have this problem I have to change the ruby version to 2.5.8. Rejected?

**Comment by deivid-rodriguez** [comment]
This issue should be independent from the ruby version in use I believe, and was fixed in bundler 2.2.3. Make sure you're using a recent enough version of bundler, and if you're still having problems, please open a separate ticket. Thanks!
